### PR TITLE
Make InternalKeyComparator not configurable

### PR DESCRIPTION
**Pull request by siying**
Summary:
InternalKeyComparator is an internal class which is a simple wrapper of Comparator. https://github.com/facebook/rocksdb/pull/8336 made Comparator customizeable. As a side effect, internal key comparator was made configurable too. This introduces overhead to this simple wrapper. For example, every InternalKeyComparator will have an std::vector attached to it, which consumes memory and possible allocation overhead too.
We remove InternalKeyComparator from being customizable by making InternalKeyComparator not a subclass of Comparator.

Test Plan: Run existing CI tests and make sure it doesn't fail

### CODE REVIEW

**Comment by mrambacher** [comment]
While I agree that it is less-than-ideal to have an unused vector here taking up space, I am wondering if this is the best solution or what problem is being solved here.   What is the performance difference with this change?

Many of the methods that are part of the Comparator have not been moved to the ComparatorInterface.  Is there a reason these were skipped?

What would happen if instead the std::vector was changed into a std::unique_ptr<std::vector>, and constructors (and guards) were added that allowed for this vector to be null?  Would the equivalent performance/size gain be accomplished for all classes/implementations that chose to use that new constructor (and not have any options)?

File: table/merging_iterator.cc [hunk 1]
```diff
@@ -386,7 +386,7 @@ void MergingIterator::SwitchToForward() {
       if (child.status() == Status::TryAgain()) {
         continue;
       }
-      if (child.Valid() && comparator_->Equal(target, child.key())) {
+      if (child.Valid() && comparator_->Compare(target, child.key()) == 0) {
```
[the remark this answers is not in the document]
Comment (siying) [reply table/merging_iterator.cc:389]: Potentially yes but now InternalKeyComparator just uses a default implementation of Equal() which does exactly what is done here. Improving efficiency of InternalKeyComparator::Equal() sounds outside the scope of the PR and I don't see a need to mix the two.

File: db/dbformat.h [hunk 2]
```diff
@@ -273,7 +277,7 @@ class InternalKeyComparator
   // value `kDisableGlobalSequenceNumber`.
   int Compare(const Slice& a, SequenceNumber a_global_seqno, const Slice& b,
               SequenceNumber b_global_seqno) const;
-  virtual const Comparator* GetRootComparator() const override {
+  virtual const Comparator* GetRootComparator() const {
```
[the remark this answers is not in the document]
Comment (siying) [reply db/dbformat.h:280]: Let me see whether I can delete the function.

File: db/dbformat.h [hunk 3]
```diff
@@ -249,17 +249,21 @@ class InternalKeyComparator
...
+  }
+
   // Same as Compare except that it excludes the value type from comparison
-  virtual int CompareKeySeq(const Slice& a, const Slice& b) const;
-  virtual void FindShortestSeparator(std::string* start,
-                                     const Slice& limit) const override;
-  virtual void FindShortSuccessor(std::string* key) const override;
+  int CompareKeySeq(const Slice& a, const Slice& b) const;
+  void FindShortestSeparator(std::string* start, const Slice& limit) const;
+  void FindShortSuccessor(std::string* key) const;
```
[the remark this answers is not in the document]
Comment (siying) [reply db/dbformat.h:266]: These functions are still used index_builder.h. Let me see whether I can move it to clear InternalKeyComparator.

File: db/dbformat.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (siying) [reply db/dbformat.h]: Removed it.

File: db/dbformat.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (siying) [reply db/dbformat.h]: Moved the functions to IndexBuilder.

File: db/dbformat_test.cc [hunk 4]
```diff
@@ -9,6 +9,7 @@
 
 #include "db/dbformat.h"
 
+#include "table/block_based/index_builder.h"
```
Comment (mrambacher) [inline db/dbformat_test.cc:12]: Nit: This should probably include compaator.h as well.

**Review by mrambacher — APPROVED** [review]
Thanks for the changes

**Pull request closed without merging**